src/library_wasmfs_fetch.js

```diff
@@ -0,0 +1,63 @@
+mergeInto(LibraryManager.library, {
+  // Fetch backend: On first access of the file (either a read or a getSize), it
+  // will fetch() the data from the network asynchronously. Otherwise, after
+  // that fetch it behaves just like JSFile (and it reuses the code from there).
+
+  _wasmfs_create_fetch_backend_js__deps: [
+    '$wasmFS$backends',
+    '$wasmFS$JSMemoryFiles',
+    '_wasmfs_create_js_file_backend_js',
+  ],
+  _wasmfs_create_fetch_backend_js: async function(backend) {
+    // Get a promise that fetches the data and stores it in JS memory (if it has
+    // not already been fetched).
+    async function getFile(file) {
+      if (wasmFS$JSMemoryFiles[file]) {
+        // The data is already here, so nothing to do before we continue on to
+        // the actual read below.
+        return Promise.resolve();
+      }
+
+      // This is the first time we want the file's data.
+      // TODO: real URL!
+      var url = 'data.dat';
+      var response = await fetch(url);
+      var buffer = await response['arrayBuffer']();
+      wasmFS$JSMemoryFiles[file] = new Uint8Array(buffer);
+    }
+
+    // Start with the normal JSFile operations. This sets
+    //   wasmFS$backends[backend]
+    // which we will then augment.
+    __wasmfs_create_js_file_backend_js(backend);
+
+    // Add the async operations on top.
+    var jsFileOps = wasmFS$backends[backend];
```

Comment:
Can we have `__wasmfs_create_js_file_backend_js(backend)` return the backend object to modify so we don't need to access the `wasmFS$backends` global state later?

Also, is this really the function we want to call here? IIUC, `__wasmfs_create_js_file_backend_js` sets up the JS for the normal JS backend, which we immediately overwrite. Would it make sense to have a `__wasmfs_register_backend` function that just registers the backend and returns a mostly-empty object to modify?

Reply:
> Can we have __wasmfs_create_js_file_backend_js(backend) return the backend object to modify so we don't need to access the wasmFS$backends global state later?

Perhaps... but it's called by C, so C obviously could not use the return value if it were an object. Atm there is no return value, so that would technically work. But it is a little confusing, and we might want to add an error code as a return value some day which would conflict.

Not sure what you're suggesting in paragraph 2. I don't see the difference between the two options you refer to there? Maybe you can sketch it out in code?

Reply:
Oh, I slightly misunderstood what was happening here. I'm sad we can't call a backend registry function from C and get an anyref back, then pass it to a separate initialization function. If these `create_X_backend_js` functions could just take, modify, and return JS backend objects, it would be much nicer to chain them together starting with the registry function that gets an empty backend object. And that would work from either C or JS.

But anyway this seems like the best we can do for now given that `_wasmfs_create_js_file_backend_js` has to be called from C as well.